networkit/cpp/graph/GraphTools.cpp

```diff
@@ -201,14 +201,65 @@ double density(const Graph &G) noexcept {
 }
 
 double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+    auto volume = 0.0;
 
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+    if (G.isWeighted()) {
+        volume = G.totalEdgeWeight();
+    } else {
+        volume = G.numberOfEdges();
+    }
+    if (!G.isDirected()) {
+        volume *= 2.0;
+    }
+    return volume;
+}
+
+double subgraphInVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool countSelfLoopsTwice = !G.isDirected();
+    count nodesNotInGraph = 0;
+
+    // Check for nodes in subset, which are not part of the graph
+    for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+        if (!G.hasNode(*it)) {
+            nodesNotInGraph++;
+        } 
+    } 
+
+    // Compute volume of subgraph if the subset is really a subset
+    if (!nodes.empty() && (G.numberOfNodes() - nodesNotInGraph) != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            inVolume += G.weightedDegreeIn(*it, countSelfLoopsTwice);
+        }
+    // Otherwise compute volume of the complete graph
+    } else {
+        inVolume = volume(G);
+    }
+    return inVolume;
+}
+
+double subgraphVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool countSelfLoopsTwice = !G.isDirected();
+    count nodesNotInGraph = 0;
+
+    // Check for nodes in subset, which are not part of the graph
+    for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+        if (!G.hasNode(*it)) {
+            nodesNotInGraph++;
+        } 
+    } 
+
+    // Compute volume of subgraph if node-subset is non-empty
+    if (!nodes.empty() > 0 && (G.numberOfNodes() - nodesNotInGraph) != nodes.size()) {
```

Comment:
The `> 0` in the first condition should be removed. The `nodesNotInGraph` should be subtracted from `nodes.size()`, not `G.numberOfNodes()`.